File: trunk/NowPlaying/iPhone/Classes/LocationUtilities.m
// Copyright 2008 Cyrus Najmabadi
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "LocationUtilities.h"

#import "Location.h"
#import "NetworkUtilities.h"
#import "Utilities.h"
#import "XmlElement.h"

@implementation LocationUtilities

+ (Location*) findLocationWithGeonames:(CLLocation*) location {
    CLLocationCoordinate2D coordinates = location.coordinate;
    double latitude = coordinates.latitude;
    double longitude = coordinates.longitude;
    NSString* url = [NSString stringWithFormat:@"http://ws5.geonames.org/findNearbyPostalCodes?lat=%f&lng=%f&maxRows=1", latitude, longitude];

    XmlElement* geonamesElement = [NetworkUtilities xmlWithContentsOfAddress:url important:YES];
    XmlElement* codeElement = [geonamesElement element:@"code"];
    NSString* postalCode = [codeElement element:@"postalcode"].text;
    NSString* country = [codeElement element:@"countryCode"].text;

    if (postalCode.length == 0) {
        return nil;
    }

    if ([@"CA" isEqual:country]) {
        return nil;
    }

    NSString* city = [codeElement element:@"adminName1"].text;
    NSString* state = [codeElement element:@"adminCode1"].text;

    return [Location locationWithLatitude:latitude
                                longitude:longitude
                                  address:@""
                                     city:city
                                    state:state
                               postalCode:postalCode
                                  country:country];
}


+ (Location*) findLocationWithGeocoder:(CLLocation*) location {
    CLLocationCoordinate2D coordinates = location.coordinate;
    double latitude = coordinates.latitude;
    double longitude = coordinates.longitude;
    NSString* url = [NSString stringWithFormat:@"http://geocoder.ca/?latt=%f&longt=%f&geoit=xml&reverse=Reverse+GeoCode+it", latitude, longitude];

    XmlElement* geodataElement = [NetworkUtilities xmlWithContentsOfAddress:url
                                                                  important:YES];
    NSString* postalCode = [geodataElement element:@"postal"].text;
    if (postalCode.length == 0) {
        return nil;
    }

    NSString* city = [geodataElement element:@"city"].text;
    NSString* state = [geodataElement element:@"prov"].text;

    return [Location locationWithLatitude:latitude
                                longitude:longitude
                                  address:@""
                                     city:city
                                    state:state
                               postalCode:postalCode
                                  country:@"CA"];
}


+ (Location*) findLocationWithGoogle:(CLLocation*) location {
    CLLocationCoordinate2D coordinates = location.coordinate;
    double latitude = coordinates.latitude;
    double longitude = coordinates.longitude;
    NSString* url = [NSString stringWithFormat:@"http://maps.google.com/maps/geo?q=%f,%f&output=xml&oe=utf8&sensor=false&key=ABQIAAAAE33gn89pf9QC1N10Oi1IxBTjs0lgCCfZJx1z0ucxfREoQjAihRQgAaDiNU3GwvKqQjMaH59qEdSkAg", latitude, longitude];
    
    XmlElement* kmlElement = [NetworkUtilities xmlWithContentsOfAddress:url important:YES];
    
    NSString* postalCode = [[kmlElement element:@"PostalCodeNumber" recurse:YES] text];
    
    if (postalCode.length == 0) {
        return nil;
    }
    
    NSString* country = [[kmlElement element:@"CountryNameCode" recurse:YES] text];
    NSString* state = [[kmlElement element:@"AdministrativeAreaName" recurse:YES] text];
    NSString* city = [[kmlElement element:@"SubAdministrativeAreaName" recurse:YES] text];
    NSString* locality = [[kmlElement element:@"LocalityName" recurse:YES] text];
    
    return  [Location locationWithLatitude:latitude
                                 longitude:longitude
                                   address:@""
                                      city:city.length > 0 ? city : locality
                                     state:state
                                postalCode:postalCode
                                   country:country];
}


+ (Location*) findLocation:(CLLocation*) location {
    Location* result = [self findLocationWithGoogle:location];
    if (result != nil) {
        return result;
    }
    
    result = [self findLocationWithGeonames:location];
    if (result != nil) {
        return result;
    }
    
    result = [self findLocationWithGeocoder:location];
    if (result != nil) {
        return result;
    }
    
    return nil;
}

@end

File: tags/NowPlaying/Android/1.3/src/org/metasyntactic/utilities/LocationUtilities.java
//Copyright 2008 Cyrus Najmabadi
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.
package org.metasyntactic.utilities;

import org.metasyntactic.data.Location;
import static org.metasyntactic.utilities.StringUtilities.isNullOrEmpty;
import static org.metasyntactic.utilities.XmlUtilities.element;
import static org.metasyntactic.utilities.XmlUtilities.text;
import org.w3c.dom.Element;

public class LocationUtilities {
  private LocationUtilities() {
  }

  private static Location findLocationWithGeonames(final double latitude, final double longitude) {
    final String url = "http://ws.geonames.org/findNearbyPostalCodes?lat=" + latitude + "&lng=" + longitude + "&maxRows=1";

    final Element geonamesElement = NetworkUtilities.downloadXml(url, true);
    final Element codeElement = element(geonamesElement, "code");
    final String postalCode = text(element(codeElement, "postalcode"));
    final String country = text(element(codeElement, "countryCode"));

    if (isNullOrEmpty(postalCode)) {
      return null;
    }

    if ("CA".equals(country)) {
      return null;
    }

    final String city = text(element(codeElement, "adminName1"));
    final String state = text(element(codeElement, "adminCode1"));

    return new Location(latitude, longitude, "", city, state, postalCode, country);
  }

  private static Location findLocationWithGeocoder(final double latitude, final double longitude) {
    final String url = "http://geocoder.ca/?latt=" + latitude + "&longt=" + longitude + "&geoit=xml&reverse=Reverse+GeoCode+it";

    final Element geodataElement = NetworkUtilities.downloadXml(url, true);
    final String postalCode = text(element(geodataElement, "postal"));

    if (isNullOrEmpty(postalCode)) {
      return null;
    }

    final String city = text(element(geodataElement, "city"));
    final String state = text(element(geodataElement, "prov"));

    return new Location(latitude, longitude, "", city, state, postalCode, "CA");
  }

  private static Location findLocationWithGoogle(final double latitude, final double longitude) {
    final String url = "http://maps.google.com/maps/geo?q=" + latitude + ',' + longitude
    + "&output=xml&oe=utf8&sensor=false&key=ABQIAAAAE33gn89pf9QC1N10Oi1IxBTjs0lgCCfZJx1z0ucxfREoQjAihRQgAaDiNU3GwvKqQjMaH59qEdSkAg";

    final Element kmlElement = NetworkUtilities.downloadXml(url, true);

    final String postalCode = text(element(kmlElement, "PostalCodeNumber", true));

    if (isNullOrEmpty(postalCode)) {
      return null;
    }

    final String country = text(element(kmlElement, "CountryNameCode", true));
    final String state = text(element(kmlElement, "AdministrativeAreaName", true));
    final String city = text(element(kmlElement, "SubAdministrativeAreaName", true));
    final String locality = text(element(kmlElement, "LocalityName", true));

    return new Location(latitude, longitude, "", isNullOrEmpty(city) ? locality : city, state, postalCode, country);
  }

  public static Location findLocation(final double latitude, final double longitude) {
    Location result = findLocationWithGoogle(latitude, longitude);
    if (result != null) {
      return result;
    }

    result = findLocationWithGeonames(latitude, longitude);
    if (result != null) {
      return result;
    }

    result = findLocationWithGeocoder(latitude, longitude);
    if (result != null) {
      return result;
    }

    return null;
  }
}
